rfcs/rfc-0003-discriminators.md

```diff
@@ -0,0 +1,229 @@
+| Unique cryptographic user identifiers with display names                | Resort to using friend codes, invite links, and the like exclusively.                                                 |             ⚠️              |           ✅            |   ⚠️    |    ⚠️     |   Medium   |
+
+In regards to this table,
+
+- Privacy means whether users can avoid being discovered based on their common names.
+- Usability is whether users can practically and quickly add each other.
+- † Restricting existing usernames further would potentially require some users to change username.
+- ¶ Removing usernames altogether would require a complete redesign of how friends work on Revolt.
+
+Discriminators also help solve a couple of issues with regards to username abandonment, users may sign up and:
+
+1. Forget their account credentials or otherwise get locked out, and never recover their account.
+2. Forget that their account even exists.
+3. Decide they don't want to keep using their account but never delete it.
+
+Revolt does not currently automatically delete old accounts with no activity which is why this poses an issue.
+
+Given Revolt's current size, 4-digit numeric discriminators currently pose the least issues.
+
+| Solution       |      Example       | Description                                   | Usability | Quantity | Safe |
+| -------------- | :----------------: | --------------------------------------------- | :-------: | :------: | :--: |
+| 4-digit        |      `#1234`       | Any 4 digits                                  |    ✅     |   Low    |  ✅  |
+| N-digit        |     `#123456`      | Any N digits                                  |    ⚠️     |  Medium  |  ⚠️  |
+| Variable Digit | `#123` ... `#1234` | Scale between n and N digits depending on use |    ⚠️     | Infinite |  ⚠️  |
+| 4-hex          |      `#12AF`       | Any 4 hex characters                          |    ✅     |  Medium  |  ⚠️  |
+| 4-char         |      `#1ABZ`       | Any 4 alphanumeric characters                 |   ❌ †    |   High   |  ❌  |
+
+In regards to this table:
+
+- Usability is whether the solution is practical, i.e. reasonably sized and simple.
+- Quantity is how many possible discriminators may be housed under one username.
+- Safe is whether the solution is not susceptible to generating undesired combinations and phrases.
+- † Allowing any alphanumeric characters may cause confusion between similar charactres using certain fonts, e.g. `O` and `0`.
+
+# Prior art
+
+This has been implemented before on other platforms:
+
+- Discord's discriminator system
+- Blizzard's [BattleTag system](https://us.battle.net/support/en/article/75767)
+
+# Unresolved questions
+
+No currently unresolved questions.
```

Comment:
gonna make this a review comment just so it isn't forgotten.

will users be able to customize their discriminators? there are very good reasons for *and* against this.

pros:
- allows full customization of how your username is presented on the platform
- instantly familiar to long-time users of discord (and can even become somewhat of a selling point, given recent events lol)
- users would (usually) be able to set their discrims to a number they'd be more likely to memorize
- people would not have to abuse the API to "re-roll" their discriminators

cons:
- discriminators that are deemed "valuable" would be immediately taken, possibly even sold for real money
- in addition to the above, users with "valuable" discriminators could be harassed into giving up their accounts
- users would easily be able to steal usernames+discriminators the moment the target user changes their name
- more work for platform moderators to ensure people don't use offensive usernames+discriminator combos

Reply:
Discriminators will be random for all users, no customisation besides re-rolling.